nav/coordinates.py

```diff
@@ -75,28 +84,89 @@ def finalize(self):
             self.crossings  = np.where(np.diff(detect))[0]
             self.day        = ~detect
 
-            # Place for future logic dev
-            #if self.day[0]:
-            #    for cr in self.crossings
+
+        self.mjd_crossings = np.fromiter(self.precise_crossings(), float)
         self.set_temperature()
 
-    ###
-    def calculate(self, interval):
-        # Note the crafty logic in the Observation class constructor - it's hand-made polymorphism.
-        # After update in Nov 2023, the Observation ctor can take a tuple of start and end time points.
+
+    ### ---
+    def clock(self, mjd):
+        """ This method calculates the "lunar clock" e.g. the time according
+            to 24-hour subdividion of the Lunar day.
+        """
         
+        if mjd>self.mjd[self.N-1] or mjd<self.mjd[0]: return None
+
+        if mjd<self.mjd_crossings[0]: # only one endpoint at the start, use the next cycle as estimate
+            estimate = self.mjd_crossings[1] - self.mjd_crossings[0]
+            if self.day[self.crossings[0]]:
+                return 6.0 + 12.0*(1.0-(self.mjd_crossings[0] - mjd)/estimate)
+            else:
+                return 6.0*(1.0-(self.mjd_crossings[0] - mjd)/estimate)
+
+        if mjd>self.mjd_crossings[-1]: # only one endpoint at the end, use the previous cycle as estimate
+            estimate = self.mjd_crossings[-1] - self.mjd_crossings[-2] # print(estimate)
+            if self.day[~self.crossings[-1]]: # print('day')
+                return 6.0 + 12.0*(1.0-(mjd - self.mjd_crossings[-1])/estimate)
+            else: # print('night')
+                return 6.0*(1.0-(mjd - self.mjd_crossings[0])/estimate)
+
+
+        indices = list(range(len(self.mjd_crossings)))
+        _ = indices.reverse()
+    
+        for i in indices:
+            if mjd>self.mjd_crossings[i]:
+                estimate = self.mjd_crossings[i+1] - self.mjd_crossings[i] #  print(estimate)
+                if (~self.day[self.crossings[i]]): # day, since crossing is one off by design
+                    return 6.0+12.0*(mjd-self.mjd_crossings[i])/estimate
+                else:
+                    result = 18.0 + 12.0*(mjd-self.mjd_crossings[i])/estimate
+                    if result >24.0: result-=24.0
+                    return result
+
+        return None
```

Comment:
In which case do we return None? `self.clock()` should be a well defined quantity so it should either return clock or throw an error if things are really sideways (i.e. you inputed an invalid mjd). Not return None.

Reply:
Done, replaced with an appropriate exception throw.